texastribune/data-visuals-create
# Create a manifest of graphic elements in output

### Changed
- `templates/graphic/project.config.js`, `templates/feature/project.config.js` - Adds new key for adding customization to the parsing step. For now the only one there is one to ignore Apple News for specified files.
- `templates/__common__/_package.json` - Adds puppeteer and a new `npm run parse` step
- `templates/feature/README.md`, `templates/graphic/README.md` - Adds new context for these changes
- `templates/graphic/app/templates/base.html` - New meta tags on graphics to be used internally.

### Added
- ` templates/__common__/_.npmrc`- This tells puppeteer to skip downloading chrome each time we create a new project. Saves us some storage on our machines.
- `templates/__common__/config/scripts/parse.js ` - Task runner that spins up a local server and kicks off the graphics-meta.js step.
- `templates/__common__/config/tasks/graphics-meta.js `- The base for this whole process. At a high level, this will step through the whole project /dist folder and extract all the relevant metadata info we specify. This is also where we build the JSON for Apple News and capture screenshots of graphics.


## Test steps:

1. Check out `graphic-metadata` and run `your/path/to/data-visuals-create/bin/data-visuals-create graphic my-test-graphic`
2. In that new project, add some sample HTML to the index.html. Feel free to use something from an old project or just this [test HTML](https://gist.github.com/ashley-hebler/44632a2c7eef5064b4d2b602dd100b97) I've been using.
3. Run `npm run serve` to preview the graphic. If you don't use the sample code, be sure to add a .button CTA to demonstrate how those will be handled in a screenshot.
3. Run `npm run build` then `npm run parse`

-- 🆕 steps since https://github.com/texastribune/data-visuals-create/pull/68/commits/3d8031b13408788977632dc8a05efc11c0c1af23 --
5. Confirm a **`manifest.json`** file is created. This is a list of the metatdata for each graphic.
6. Confirm a small and large preview .png is also created

## Changes to dist output after running `npm run parse`
**before:**
![Screen Shot 2021-01-27 at 3 11 36 PM](https://user-images.githubusercontent.com/2974713/106054909-47347f00-60b2-11eb-856a-4aa0c5fd8820.png)
**after:**
![Screen Shot 2021-01-27 at 3 12 29 PM](https://user-images.githubusercontent.com/2974713/106054907-456abb80-60b2-11eb-8bb4-404f690f61a8.png)




## Notes:
@mandicai - ~I didn't add the tags to the project config yet, but we can add as many new keys there as we see fit.~ 
2/4/21 Update: tags are now there. I just cherry-picked this commit https://github.com/texastribune/data-visuals-create/pull/34/commits/3a53493957c7818d4ec0dc2dec1d1e328ea32998

 I know there's quite a bit to review so take your time. I hope it's not too messy! Let me know if you want to talk through anything.

Also for now, I left this off of the build step, but it might make sense to make it more automatic and make [this line](https://github.com/texastribune/data-visuals-create/blob/a26700093e6afc0012d24d10f10b3e5c84104f1f/templates/__common__/_package.json#L7):
`"build": "NODE_ENV=production node ./config/scripts/build.js && npm run parse",`
once we verify it the process is working for us.
